src/ec/eddsa.rs

```diff
@@ -112,9 +113,52 @@ impl<'a> Ed25519KeyPair {
     /// Returns the signature of the message `msg`.
     pub fn sign(&self, msg: &[u8]) -> signature::Signature {
         let mut signature_bytes = [0u8; 64];
+        let mut az = [0u8; digest::SHA512_OUTPUT_LEN];
+        digest::GFp_SHA512_4(az.as_mut_ptr(), az.len(),
+                             self.private_public.as_ptr(), 32,
+                             core::ptr::null(), 0,
+                             core::ptr::null(), 0,
+                             core::ptr::null(), 0);
         unsafe {
-            GFp_ed25519_sign(signature_bytes.as_mut_ptr(), msg.as_ptr(),
-                             msg.len(), self.private_public.as_ptr());
+            GFp_ed25519_scalar_mask(
+                slice_as_array_ref_mut!(&mut az[..SCALAR_LEN], SCALAR_LEN).unwrap())
+        };
+        let mut nonce = [0u8; digest::SHA512_OUTPUT_LEN];
+        let az_offset_ptr = unsafe { az.as_ptr().offset(32) };
+        digest::GFp_SHA512_4(nonce.as_mut_ptr(), nonce.len(),
+                             az_offset_ptr, 32,
+                             msg.as_ptr(), msg.len(),
+                             core::ptr::null(), 0,
+                             core::ptr::null(), 0);
+        unsafe { GFp_x25519_sc_reduce(nonce.as_mut_ptr()) };
+        let mut r = Point {
+            x: [0; ELEM_LIMBS],
+            y: [0; ELEM_LIMBS],
+            z: [0; ELEM_LIMBS],
+            t: [0; ELEM_LIMBS],
+        };
+        unsafe {
+            let nonce_half_ref_mut =
+                slice_as_array_ref_mut!(&mut nonce[..SCALAR_LEN], SCALAR_LEN).unwrap();
+            GFp_x25519_ge_scalarmult_base(&mut r, nonce_half_ref_mut);
+            let signature_bytes_half_ref_mut =
+                slice_as_array_ref_mut!(&mut signature_bytes[..SCALAR_LEN], SCALAR_LEN).unwrap();
+
+            GFp_ge_p3_tobytes(signature_bytes_half_ref_mut, &r);
+        }
+        let mut hram = [0u8; digest::SHA512_OUTPUT_LEN];
+        let public_private_offset_ptr = unsafe {
+            self.private_public.as_ptr().offset(32)
+        };
```

Comment:
Same things apply here. Use `split_at_mut()` and/or slicing, use a named constant instead of `32`, use the regular Digest API.